File: app/src/main/java/com/hepta/theptavpn/LocalVPNService.java
package com.hepta.theptavpn;


import android.app.PendingIntent;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.VpnService;
import android.os.Handler;
import android.os.Looper;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import android.widget.Toast;

import com.hepta.theptavpn.vpnfirewall.PingReflector;

import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;


public class LocalVPNService extends VpnService
{


    static {
        System.loadLibrary("theptavpn");
    }
    public native boolean connect_server(String serverAddress, String serverPort);

    public native void StartVpn(int fd, int prorxType);
    public native void stopVpn();

    public native void startProxyServer();
    public static final String ACTION_DISCONNECT = "ACTION_DISCONNECT";
    private static int MTU = 1400;

    private PingReflector mPingReflector = null;
    private static final String TAG = LocalVPNService.class.getSimpleName();
//    private static final String VPN_ADDRESS = "192.168.0.101"; // Only IPv4 support for now 掩码 24
    private static final String VPN_ADDRESS = "10.0.0.2"; // Only IPv4 support for now

    //    private static final String VPN_ROUTE = "192.168.0.0"; // Intercept everything  掩码 24
    private static final String VPN_ROUTE = "0.0.0.0"; // Intercept everything
    private ParcelFileDescriptor vpnInterface = null;

    private PendingIntent pendingIntent;

    @Override
    public void onCreate()
    {
        super.onCreate();
    }

    private void setupVPN()
    {
        if (vpnInterface == null)
        {
            Builder builder = new Builder();
            builder.addAddress(VPN_ADDRESS, 30); //第二个参数子网掩码
            try {
                InetAddress address = InetAddress.getByName(VPN_ROUTE);
                builder.addRoute(address, 0); //第二个参数子网掩码

            } catch (UnknownHostException e) {
                throw new RuntimeException(e);
            }
            builder.setMtu(MTU);
            //                builder.addDisallowedApplication("com.android.chrome");  //禁止这个应用通过vpn访问网络，但是不禁止网络，就像vpn不存在一样，正常访问网络,可以设置多个
            try {
                builder.addAllowedApplication("com.hepta.vpntest");  // 只允许这个应用通过vpn访问网络，其他应用不禁止网络，就像vpn不存在一样，正常访问网络，可以设置多个
            } catch (PackageManager.NameNotFoundException e) {
                throw new RuntimeException(e);
            }
//            builder.addAllowedApplication("com.tencent.mm");  //
//            builder.allowBypass();
            vpnInterface = builder.setSession(getString(R.string.app_name)).setConfigureIntent(pendingIntent).establish();
            Log.e(TAG,"fd:"+vpnInterface.getFd());
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null && ACTION_DISCONNECT.equals(intent.getAction())) {
            disconnect();
            stopVpn();
            return START_NOT_STICKY;
        } else {
            String ServerAddress = intent.getStringExtra("serverAddress");
            String ServerPort = intent.getStringExtra("serverPort");
            setupVPN();
            StartVpn(vpnInterface.getFd(),1);
            return START_STICKY;
        }
    }

    @Override
    public boolean stopService(Intent name) {
        return super.stopService(name);
    }

    private void disconnect() {
        cleanup();
        Log.i(TAG, "Stopped");
    }

    private void connect(String serverAddress, String serverPort) {

        if(!connect_server(serverAddress,serverPort)){
            showDialog("不能连接到服务");
            return;
        };


    }

    @Override
    public void onDestroy()
    {
        super.onDestroy();

    }

    private void cleanup()
    {
        if (mPingReflector != null) {
            mPingReflector.interrupt();
            mPingReflector = null;
        }
        try {
            if (vpnInterface != null) {
                Log.i(TAG, "Closing filedescriptor");
                vpnInterface.getFd();
                vpnInterface.close();
            }
        } catch(IOException e) {
            Log.w(TAG, "Closing filedescriptor failed", e);
        } finally {
            vpnInterface = null;
        }
    }

    private void showDialog(String msg){
        Handler handlerThree=new Handler(Looper.getMainLooper());
        handlerThree.post(new Runnable(){
            public void run(){
                Toast.makeText(getApplicationContext() ,msg,Toast.LENGTH_LONG).show();
            }
        });
    }
}

File: app/src/main/java/com/hepta/theptavpn/LocalVPNService.kt
package com.hepta.theptavpn

import android.content.Intent
import android.net.VpnService
import android.os.*
import android.util.Log
import android.widget.Toast
import com.hepta.theptavpn.Tunnel.IPreflectorTunnel
import com.hepta.theptavpn.Tunnel.ProxyTunnel
import com.hepta.theptavpn.Tunnel.tun2sockTunnel
import java.net.InetAddress
import java.net.UnknownHostException


class LocalVPNService : VpnService() {


    private var proxyBinder = LocalVPNService.ProxyBinder(this@LocalVPNService)
    private var tunnel: ProxyTunnel? =null
    public var Running: Boolean?=false
    private fun setupVPN(): ParcelFileDescriptor? {
        val ipaddr = MmkvManager.getAddress();
        val mtu = MmkvManager.getMtu();
        val netmask = MmkvManager.getnetMask();
        val prefixLength = Utils.netmask2prefix(netmask)
        val builder: Builder = Builder()
        builder.setMtu(mtu)
        builder.addAddress(ipaddr, prefixLength) //第二个参数子网掩码
        try {
            val address = InetAddress.getByName(VPN_ROUTE)
            builder.addRoute(address, 0) //第二个参数子网掩码
        } catch (e: UnknownHostException) {
            throw RuntimeException(e)
        }
        val allow_type = MmkvManager.getAllowType()
        when (allow_type) {
            MmkvManager.KEY_APP_ALLWO_NONE -> {
                builder.addDisallowedApplication(packageName)
            }
            MmkvManager.KEY_APP_ADD_ALLOW -> {
                builder.addAllowedApplication("")
                val applist = MmkvManager.decodeApplicationList(MmkvManager.KEY_APP_ADD_ALLOW)
                for (appPkg in applist) {
                    builder.addAllowedApplication(appPkg)
                }
            }
            MmkvManager.KEY_APP_ADD_DIS_ALLOW -> {
                builder.addDisallowedApplication(packageName)
                val applist = MmkvManager.decodeApplicationList(MmkvManager.KEY_APP_ADD_DIS_ALLOW)
                for (appPkg in applist) {
                    builder.addDisallowedApplication(appPkg)
                }
            }
        }
        return builder.setSession(getString(R.string.app_name)).establish()

    }
    private fun getTunnelType(config: ServerConfig, parcelFileDescriptor: ParcelFileDescriptor): ProxyTunnel {

        if (config.netType == 0) {
            return IPreflectorTunnel(config, parcelFileDescriptor)
        } else {
            return tun2sockTunnel(config, parcelFileDescriptor)
        }
    }

    public fun stopVpnService(){
        tunnel?.stop()
        proxyBinder.updateRunStatus(false);
    }

    public fun startVpnService(guid:String){
        val config = MmkvManager.decodeServerConfig(guid)
        setupVPN()?.let {
            tunnel = getTunnelType(config!!, it)
            val run = tunnel?.start();
            proxyBinder.updateRunStatus(run!!);
            if(!run){
                Log.e("Rzx","start vpn failed");
            }
        }
    }

    private fun showDialog(msg: String) {
        val handlerThree = Handler(Looper.getMainLooper())
        handlerThree.post { Toast.makeText(applicationContext, msg, Toast.LENGTH_LONG).show() }
    }

    public override fun onRevoke() {
        Log.e("Rzx","onRevoke")
        stopVpnService()
        proxyBinder.updateRunStatus(false);
        super.onRevoke()

    }
    public override fun onBind(intent: Intent) : IBinder {
        //前面这几行很重要，否则无法调用 onRevoke
        val action = if (intent != null) intent.action else null
        if (action != null && action == SERVICE_INTERFACE) {
            return super.onBind(intent)!!
        }
        return proxyBinder
    }

    public class ProxyBinder(private val localVPNService: LocalVPNService) : Binder(){

        private var mainActivity : MainActivity?= null
        public fun getLocalVPNService(activity: MainActivity) : LocalVPNService{
            mainActivity = activity
            return localVPNService
        }

        public fun updateRunStatus( status:Boolean){
            mainActivity?.mainViewModel?.isRunning?.postValue(status)
            localVPNService.Running=status
        }

    }

    companion object {
        init {
            System.loadLibrary("theptavpn")
        }
        private val TAG = LocalVPNService::class.java.simpleName
        private const val VPN_ROUTE = "0.0.0.0" // Intercept everything

    }
}